numpy/lib/format.py

```diff
@@ -304,27 +305,32 @@ def _write_array_header(fp, d, version=None):
         header.append("'%s': %s, " % (key, repr(value)))
     header.append("}")
     header = "".join(header)
-    # Pad the header with spaces and a final newline such that the magic
-    # string, the header-length short and the header are aligned on a
-    # 16-byte boundary.  Hopefully, some system, possibly memory-mapping,
-    # can take advantage of our premature optimization.
-    current_header_len = MAGIC_LEN + 2 + len(header) + 1  # 1 for the newline
-    topad = 16 - (current_header_len % 16)
-    header = header + ' '*topad + '\n'
     header = asbytes(_filter_header(header))
 
-    hlen = len(header)
-    if hlen < 256*256 and version in (None, (1, 0)):
+    hlen = len(header) + 1 # 1 for newline
+    padlen1 = ARRAY_ALIGN - ((MAGIC_LEN + struct.calcsize('<H') + hlen) % ARRAY_ALIGN)
+    padlen2 = ARRAY_ALIGN - ((MAGIC_LEN + struct.calcsize('<I') + hlen) % ARRAY_ALIGN)
+
+    if hlen + padlen1 < 2**16 and version in (None, (1, 0)):
```

Comment:
We could, but it would turn a 1-deep if/else into a N-deep try/except chain where N is the number of versions.  And we definitely know in advance whether the pack would fail, so in my opinion it is better style to check it explicitly.  The [format documentation](https://docs.scipy.org/doc/numpy/neps/npy-format.html) explicitly mentions 65535 as the maximum size; I'd be happy to use that decimal form instead of 2**16 if you prefer.